In [1]:
from transformers import TFBertModel, AutoTokenizer
from scipy.spatial.distance import cosine

2025-04-28 23:14:45.077965: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  SSE4.1 SSE4.2
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.


# Model Setup
Load the pre-trained BERT model and tokenizer. This model will help us extract embeddings for our analysis

In [2]:
# Defining the model name
model_name = "bert-base-cased"

# Loading the pre-trained model and tokenizer
model = TFBertModel.from_pretrained(model_name)
tokenizer = AutoTokenizer.from_pretrained(model_name)

2025-04-28 23:15:50.349519: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  SSE4.1 SSE4.2
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
Some layers from the model checkpoint at bert-base-cased were not used when initializing TFBertModel: ['nsp___cls', 'mlm___cls']
- This IS expected if you are initializing TFBertModel from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing TFBertModel from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
All the layers of TFBertModel were initialized from the model check

# Function Definition: Predict
Define a function that encodes input text into tensors, which are then fed to the model to obtain embeddings

In [3]:
# Defining a function to encode the input text and get model predictions
def predict(text):
    encoded_inputs = tokenizer(text, return_tensors="tf")
    return model(**encoded_inputs)[0]

# Defining the sentences
Set up sentences to analyze

In [4]:
#Example 1:
# Defining the sentences
sentence1 = "There was a fly drinking from my soup"
sentence2 = "There was a fly drinking from my soup"
# sentence2 = "There is a fly swimming in my juice"
# sentence2 = "To become a commercial pilot, he had to fly for 1500 hours." #second fly example

# Tokenizing the sentences
tokens1 = tokenizer.tokenize(sentence1)
tokens2 = tokenizer.tokenize(sentence2)

# Tokenization and Model Predictions
Tokenize the sentences and obtain predictions (embeddings) from the model.

In [5]:
# Getting model predictions for the sentences
out1 = predict(sentence1)
out2 = predict(sentence2)

# Extracting Embeddings
Extract embeddings specifically for the word "fly" from both sentences.

In [6]:
# Extracting embeddings for the word 'fly' in both sentences
emb1 = out1[0:, tokens1.index("fly"), :][0]
emb2 = out2[0:, tokens2.index("fly"), :][0]

# Calculating the Cosine Similarity
Calculate the cosine similarity between the embeddings of the word "fly" from both sentences to measure how context affects meaning.

In [7]:
# calculating the cosine similarity between the embeddings
cosine(emb1, emb2)

0

Here since since sentence1 and sentence2 has the same exact text, cosine similarity of the token "fly" is 0.

Next: let's try another example for "fly"

In [12]:
#Example 2: 
# Defining the sentences
sentence1 = "There was a fly drinking from my soup"
sentence2 = "There is a fly swimming in my juice"

# Tokenizing the sentences
tokens1 = tokenizer.tokenize(sentence1)
tokens2 = tokenizer.tokenize(sentence2)

In [13]:
# Getting model predictions for the sentences
out1 = predict(sentence1)
out2 = predict(sentence2)

In [14]:
# Extracting embeddings for the word 'fly' in both sentences
emb1 = out1[0:, tokens1.index("fly"), :][0]
emb2 = out2[0:, tokens2.index("fly"), :][0]

In [15]:
# calculating the cosine similarity between the embeddings
cosine(emb1, emb2)

0.06798797845840454

In [ ]:
Here since since sentence1 and sentence2 are different, but fly word has mostly same context,
cosine similarity of the token "fly" here is 0.06798797845840454.

Eventhough, here 'fly' has mostly the same context but since they are different sentences,
the same token 'fly' has different embedding, 
How do we know ? Because the cosine similariy is not 0 here, unlike the first example.
Next: let's try one more example for "fly"

In [16]:
#Example 3:
# Defining the sentences
sentence1 = "There was a fly drinking from my soup"
sentence2 = "To become a commercial pilot, he had to fly for 1500 hours." #second fly example

# Tokenizing the sentences
tokens1 = tokenizer.tokenize(sentence1)
tokens2 = tokenizer.tokenize(sentence2)

In [17]:
# Getting model predictions for the sentences
out1 = predict(sentence1)
out2 = predict(sentence2)

In [18]:
# Extracting embeddings for the word 'fly' in both sentences
emb1 = out1[0:, tokens1.index("fly"), :][0]
emb2 = out2[0:, tokens2.index("fly"), :][0]

In [19]:
# calculating the cosine similarity between the embeddings
cosine(emb1, emb2)

0.40477824211120605

Here since since sentence1 and sentence2 have completely different context,
cosine similarity of the token "fly" here is 0.40477824211120605, which is much higher than the
previous two examples

so, eventhough the token is same i.e 'fly', it's embedding is formed according to the context, which is also called contextual embedding that captures the meaning within the 
context of the input text.

That's why in the third example the cosine similarity value is much higher than the second example.

# Conclusion
Here we saw the process of extracting and comparing word embeddings using BERT. Such techniques are fundamental in understanding word semantics and their usage across contexts.